# Kosmetyki męskie - część I - Preprocessing

### Wstęp

Prezentowane na tej stronie materiały są zapisem drugiego etapu projektu "Kosmetyki męskie". 

Opis projektu: 

I etap (pobranie danych):

Dane pobrane w pierwszym etapie projektu zostały zapisane jako *Data Frame* w pliku .pkl (*pickle*).

Celem tego etapu jest przygotowanie danych (sprawdzenie integralności, czyszczenie, konwersja).

#### Pobranie niezbędnych bibliotek

In [44]:
import pandas as pd

#### Podstawowe informacje o danych

In [45]:
df = pd.read_pickle("../data/male_products_data.pkl")
print(f'Dane posiadają {df.shape[0]} wierszy i {df.shape[1]} kolumn.')

Dane posiadają 1794 wierszy i 19 kolumn.


In [46]:
df.head(3)

product_id                      name  \
0    2106072         RITUALS... Homme    
1    2095846  WILKINSON SWORD Hydro 5    
2    2083791       JAGUAR Era Reloaded   

                                             caption                  cat_1  \
0                         Balsam do stylizacji brody  Pielęgnacja i higiena   
1             Maszynka do golenia rączka + 9 wkładów  Pielęgnacja i higiena   
2  Zestaw dla mężczyzn, woda toaletowa 100ml + że...              Mężczyzna   

            cat_2                        cat_3                      cat_4  \
0  Broda i zarost  Mycie i pielęgnacja zarostu  Balsamy i olejki do brody   
1  Broda i zarost              Golenie zarostu        Maszynki tradycyjne   
2  Perfumy męskie            Zestawy zapachowe                              

  cat_5                                                url capacity  \
0        https://www.rossmann.pl/Produkt/Balsamy-i-olej...    30 ml   
1        https://www.rossmann.pl/Produkt/Maszynki-trady...   1 szt.   
2        https://www.rossmann.pl/Produkt/Zestawy-zapach...   1 szt.   

   current_price_pln regular_price_pln lowest_price_last_30_days_pln  \
0              74.99                                           93.99   
1              79.99                                          119.99   
2             154.99                                                   

   price_per_unit_pln    unit  mega_badge  \
0              249.97  100 ml           1   
1               79.99  1 szt.           1   
2              154.99  1 szt.           0   

                                         ingredients rating reviews  
0  Ingredients: Ingredients: Aqua/Water, Hydrogen...                 
1  PEG-115M, Polyquaternium-10, Silica [Nano], Aq...                 
2  Alcohol Denat., Parfum (Fragrance), Aqua (Wate...    4.5       2

#### Sprawdzenie czy występują w danych jakieś puste kolumny i wiersze.

In [47]:
# Czy występują puste koloumny?
df.isna().all()

product_id                       False
name                             False
caption                          False
cat_1                            False
cat_2                            False
cat_3                            False
cat_4                            False
cat_5                            False
url                              False
capacity                         False
current_price_pln                False
regular_price_pln                False
lowest_price_last_30_days_pln    False
price_per_unit_pln               False
unit                             False
mega_badge                       False
ingredients                      False
rating                           False
reviews                          False
dtype: bool

In [48]:
# Czy występują puste wiersze?
empty_rows = df[df.isna().all(axis=1)]
# Wypisuje listę pustych indeksów
print(empty_rows.index.tolist())

[]


#### Sprawdzenie i konwersja typów danych.

In [49]:
# Typy danych
df.dtypes

product_id                        object
name                              object
caption                           object
cat_1                             object
cat_2                             object
cat_3                             object
cat_4                             object
cat_5                             object
url                               object
capacity                          object
current_price_pln                float64
regular_price_pln                 object
lowest_price_last_30_days_pln     object
price_per_unit_pln               float64
unit                              object
mega_badge                         int64
ingredients                       object
rating                            object
reviews                           object
dtype: object

In [50]:
# Konwersja typów danych
df['product_id'] = df['product_id'].astype(int)
df['name'] = df['name'].astype(pd.StringDtype())
df['caption'] = df['caption'].astype(pd.StringDtype())
df['cat_1'] = df['cat_1'].astype(pd.StringDtype())
df['cat_2'] = df['cat_2'].astype(pd.StringDtype())
df['cat_3'] = df['cat_3'].astype(pd.StringDtype())
df['cat_4'] = df['cat_4'].astype(pd.StringDtype())
df['cat_5'] = df['cat_5'].astype(pd.StringDtype())
df['url'] = df['url'].astype(pd.StringDtype())
df['capacity'] = df['capacity'].astype(pd.StringDtype())
df['unit'] = df['unit'].astype(pd.StringDtype())
df['ingredients'] = df['ingredients'].astype(pd.StringDtype())
df["regular_price_pln"] = pd.to_numeric(df["regular_price_pln"], errors="coerce")
df["lowest_price_last_30_days_pln"] = pd.to_numeric(df["lowest_price_last_30_days_pln"], errors="coerce")
df['mega_badge'] = df['mega_badge'].astype(bool)
df["rating"] = pd.to_numeric(df["rating"], errors="coerce")
df["reviews"] = pd.to_numeric(df["reviews"], errors="coerce")

df.dtypes

product_id                                int64
name                             string[python]
caption                          string[python]
cat_1                            string[python]
cat_2                            string[python]
cat_3                            string[python]
cat_4                            string[python]
cat_5                            string[python]
url                              string[python]
capacity                         string[python]
current_price_pln                       float64
regular_price_pln                       float64
lowest_price_last_30_days_pln           float64
price_per_unit_pln                      float64
unit                             string[python]
mega_badge                                 bool
ingredients                      string[python]
rating                                  float64
reviews                                 float64
dtype: object

#### Sprawdzenie unikalności kolumny product_id

In [51]:
df['product_id'].is_unique

True

#### Rozdzielenie wartości i jednostek w kolumnach *capacity* i *unit*
Podział na capacity_value, capacity_unit, price_reference_value, price_reference_unit a price_per_unit_pln zamieniam na price_reference_pln

In [52]:
df['capacity_value'] = df['capacity'].str.extract(r'(\d+)').astype(int)
df['capacity_unit'] = df['capacity'].str.extract(r'([a-zA-ZżźćńółęąśŻŹĆĄŚĘŁÓŃ]+)')
df['price_reference_value'] = df['unit'].str.extract(r'(\d+)').astype(int)
df['price_reference_unit'] = df['unit'].str.extract(r'([a-zA-ZżźćńółęąśŻŹĆĄŚĘŁÓŃ]+)')
df['price_reference_pln'] = df['price_per_unit_pln']

df = df.drop(columns=['capacity', 'unit', 'price_per_unit_pln'])

df.head(3)

product_id                      name  \
0     2106072         RITUALS... Homme    
1     2095846  WILKINSON SWORD Hydro 5    
2     2083791       JAGUAR Era Reloaded   

                                             caption                  cat_1  \
0                         Balsam do stylizacji brody  Pielęgnacja i higiena   
1             Maszynka do golenia rączka + 9 wkładów  Pielęgnacja i higiena   
2  Zestaw dla mężczyzn, woda toaletowa 100ml + że...              Mężczyzna   

            cat_2                        cat_3                      cat_4  \
0  Broda i zarost  Mycie i pielęgnacja zarostu  Balsamy i olejki do brody   
1  Broda i zarost              Golenie zarostu        Maszynki tradycyjne   
2  Perfumy męskie            Zestawy zapachowe                              

  cat_5                                                url  current_price_pln  \
0        https://www.rossmann.pl/Produkt/Balsamy-i-olej...              74.99   
1        https://www.rossmann.pl/Produkt/Maszynki-trady...              79.99   
2        https://www.rossmann.pl/Produkt/Zestawy-zapach...             154.99   

   ...  lowest_price_last_30_days_pln  mega_badge  \
0  ...                          93.99        True   
1  ...                         119.99        True   
2  ...                            NaN       False   

                                         ingredients rating  reviews  \
0  Ingredients: Ingredients: Aqua/Water, Hydrogen...    NaN      NaN   
1  PEG-115M, Polyquaternium-10, Silica [Nano], Aq...    NaN      NaN   
2  Alcohol Denat., Parfum (Fragrance), Aqua (Wate...    4.5      2.0   

   capacity_value  capacity_unit price_reference_value  price_reference_unit  \
0              30             ml                   100                    ml   
1               1            szt                     1                   szt   
2               1            szt                     1                   szt   

  price_reference_pln  
0              249.97  
1               79.99  
2              154.99  

[3 rows x 21 columns]

#### Weryfikacja wartości w kolumnach *capacity_unit* i *price_reference_unit*

In [53]:
df['capacity_unit'].value_counts()

capacity_unit
ml        1374
szt        315
para        62
g           27
pary         9
x            6
gratis       1
Name: count, dtype: Int64

In [54]:
df[df['capacity_unit'].isin(['x', 'gratis'])]

product_id                                         name  \
202      2111090                             GILLETTE Mach 3    
253      2110984                           GILLETTE Fusion 5    
258       245532                             SEBORADIN Forte    
473       108416           WILKINSON SWORD Xtreme 3 Sensitive   
570      2099093  L'BIOTICA BIOVAX Trychologic Advanced Detox   
1734     2073406                                    DERMENA     
1750     2099092                L'BIOTICA BIOVAX Trychologic    

                                                caption  \
202   Wkłady do maszynki do golenia, 3-ostrzowe, dla...   
253   Wkłady do maszynki do golenia 5-ostrzowe, dla ...   
258   Ampułki do włosów, przyspieszające odrost, unisex   
473   Maszynki do golenia, 3-ostrzowe, jednorazowe, ...   
570   Ampułki do włosów, kuracja redukująca przetłus...   
1734   Kuracja do włosów w ampułkach, przeciw wypadaniu   
1750            Ampułki do włosów, kuracja dla mężczyzn   

                      cat_1           cat_2                           cat_3  \
202   Pielęgnacja i higiena  Broda i zarost                 Golenie zarostu   
253   Pielęgnacja i higiena  Broda i zarost                 Golenie zarostu   
258                   Włosy     Skóra głowy  Serum i wcierki do skóry głowy   
473   Pielęgnacja i higiena  Broda i zarost                 Golenie zarostu   
570                   Włosy     Skóra głowy  Serum i wcierki do skóry głowy   
1734                  Włosy     Skóra głowy  Serum i wcierki do skóry głowy   
1750              Mężczyzna           Włosy         Pielęgnacja skóry głowy   

                            cat_4 cat_5  \
202   Wkłady i ostrza do maszynek         
253   Wkłady i ostrza do maszynek         
258                                       
473           Maszynki tradycyjne         
570                                       
1734                                      
1750                                      

                                                    url  current_price_pln  \
202   https://www.rossmann.pl/Produkt/Wklady-i-ostrz...             119.99   
253   https://www.rossmann.pl/Produkt/Wklady-i-ostrz...             139.99   
258   https://www.rossmann.pl/Produkt/Serum-i-wcierk...              41.99   
473   https://www.rossmann.pl/Produkt/Maszynki-trady...              15.49   
570   https://www.rossmann.pl/Produkt/Serum-i-wcierk...              79.99   
1734  https://www.rossmann.pl/Produkt/Serum-i-wcierk...              49.99   
1750  https://www.rossmann.pl/Produkt/Pielegnacja-sk...              79.99   

      ...  lowest_price_last_30_days_pln  mega_badge  \
202   ...                         169.99        True   
253   ...                         199.99        True   
258   ...                          48.99       False   
473   ...                          15.99        True   
570   ...                            NaN       False   
1734  ...                            NaN       False   
1750  ...                            NaN       False   

                                            ingredients rating  reviews  \
202   PEG-45M, PEG-180M, PEG-100, PEG-7M, Methyl Di-...    NaN      NaN   
253   PEG-115M, PEG-7M, PEG-100, Silica, Tocopheryl ...    NaN      NaN   
258   Ingredients: Aqua, Alcohol Denat., Propylene G...    5.0      3.0   
473   PEG-115M, PVP, PEG-100, Cyclodextrin, Tocopher...    4.8     43.0   
570   Ingredients: Aqua, Niacinamide, Propanediol, G...    NaN      NaN   
1734  Aqua, Alcohol, Pentylene Glycol, Isopropyl Alc...    4.8     13.0   
1750  Ingredients: Aqua, Alcohol Denat., Propylene G...    5.0      4.0   

      capacity_value  capacity_unit price_reference_value  \
202                3              x                     1   
253                3              x                     1   
258               14              x                   100   
473                6         gratis                     1   
570                7              x                   

In [55]:
unit_mapping = {
    'para': 'szt',
    'pary': 'szt',
    'x': 'szt',
    'gratis': 'szt'
}
df['capacity_unit'] = df['capacity_unit'].replace(unit_mapping)

df['capacity_unit'].value_counts()

capacity_unit
ml     1374
szt     393
g        27
Name: count, dtype: Int64

In [56]:
df['price_reference_unit'].value_counts()

price_reference_unit
ml     1378
szt     389
g        27
Name: count, dtype: Int64

#### Weryfikacja danych w kolumnach liczbowych

In [57]:
df.describe()

product_id  current_price_pln  regular_price_pln  \
count  1.794000e+03        1794.000000         171.000000   
mean   1.116550e+06         130.268094          52.965439   
std    8.985746e+05         175.204709          61.454278   
min    3.180000e+02           2.190000           2.790000   
25%    3.193058e+05          19.990000          17.490000   
50%    4.200335e+05          51.490000          24.990000   
75%    2.088604e+06         169.990000          53.990000   
max    2.114179e+06        1099.990000         349.990000   

       lowest_price_last_30_days_pln       rating      reviews  \
count                     533.000000  1059.000000  1059.000000   
mean                       97.232402     4.820019    10.140699   
std                       117.413026     0.347983    11.589874   
min                         2.190000     1.000000     1.000000   
25%                        19.990000     4.800000     2.000000   
50%                        51.990000     4.900000     6.000000   
75%                       121.990000     5.000000    14.000000   
max                      1119.990000     5.000000   105.000000   

       capacity_value  price_reference_value  price_reference_pln  
count     1794.000000            1794.000000           1794.00000  
mean        93.658305              78.533445            157.91903  
std        102.718019              40.808139            217.66709  
min          1.000000               1.000000              0.67000  
25%         30.000000             100.000000             16.99000  
50%         75.000000             100.000000             63.98000  
75%        100.000000             100.000000            229.99000  
max        900.000000             100.000000           1799.98000

**current_price_pln** - wszystkie produkty mają cenę, zakres wygląda prawidłowo ✅

**regular_price_pln** - 171 produktów posiada cenę katalogową, zakres wygląda prawidłowo ✅

**lowest_price_last_30_days_pln** - 533 produkty posiadają najniższą cenę za 30 dni, zakres wygląda prawidłowo ✅

**rating** - 1059 produktów posiada rating, dane mocno skrzywione (średnia 4.8) ✅

**reviews** - tyle samo produktów posiada oceny użytkowników, zakres wygląda ok (1-105) ✅

**capacity_value** - wszystkie produkty posiadają wielkość, zakresy ok ✅

**price_reference_value** - wszystkie produkty posiadają wielkość referencyjną, zakres ok ✅

**price_reference_pln** - wszystkie produkty posiadają cenę referencyjną, zakres ok ✅

#### Przegląd kategorii produktów

Do analizy potrzebuję produktów z kategorii kosmetyków. W kategorii mężczyzna znajdują się także inne produkty, które muszę odsiać.

In [58]:
# Przegląd kategorii pierwszego poziomu
df['cat_1'].value_counts()

cat_1
Mężczyzna                890
Pielęgnacja i higiena    367
Perfumy                  351
Lifestyle                104
Włosy                     82
Name: count, dtype: Int64

In [59]:
# Przegląd kategorii drugiego poziomu w kategorii Lifestyle
df[df['cat_1'] == 'Lifestyle']['cat_2'].value_counts()

cat_2
Odzież i dodatki    98
Podróż               6
Name: count, dtype: Int64

In [60]:
# Weryfikacja kategorii Podróż
df[df['cat_2'] == 'Podróż']['cat_3']

427     Minikosmetyki
495     Minikosmetyki
749     Minikosmetyki
1062    Minikosmetyki
1602    Minikosmetyki
1729    Minikosmetyki
Name: cat_3, dtype: string

In [61]:
# Usunięcie produktów z kategorii Lifestyle > Odzież i dodatki
df = df.query('not (cat_1 == "Lifestyle" and cat_2 == "Odzież i dodatki")')
df[df['cat_1'] == 'Lifestyle']['cat_2'].value_counts()

cat_2
Podróż    6
Name: count, dtype: Int64

In [62]:
# Przegląd kategorii drugiego poziomu w kategorii Mężczyzna
df[df['cat_1'] == 'Mężczyzna']['cat_2'].value_counts()

cat_2
Perfumy męskie    505
Higiena           107
Odzież męska      105
Pielęgnacja        90
Golenie            41
Włosy              31
Zarost             11
Name: count, dtype: Int64

In [63]:
# Usunięcie produktów w kategorii Mężczyzna > Odzież męska
df = df.query('not (cat_1 == "Mężczyzna" and cat_2 == "Odzież męska")')
df[df['cat_1'] == 'Mężczyzna']['cat_2'].value_counts()

cat_2
Perfumy męskie    505
Higiena           107
Pielęgnacja        90
Golenie            41
Włosy              31
Zarost             11
Name: count, dtype: Int64

In [64]:
# Przegląd podkategorii Mężczyzna > Zarost
df.query('cat_1 == "Mężczyzna" and cat_2 == "Zarost"')

product_id                       name  \
236      2085517  EWA SCHMITT Beard Master    
325       270396                ISANA MEN     
418      2078192              AA MEN Beard    
572      2085518                  BARBERO     
711       339519                  BARBERO     
744       267581       TOŁPA Dermo Barber.    
759       419910     HAGI Men Whisky Barber   
1185      270391                ISANA MEN     
1291      396104       ARGANOVE Cool Exotic   
1370     2100676          RITUALS... Homme    
1582     2100671          RITUALS... Homme    

                                                caption      cat_1   cat_2  \
236                                   Szczotka do brody  Mężczyzna  Zarost   
325   Pomada do brody, z Olejem Arganowym i Woskiem ...  Mężczyzna  Zarost   
418                  Koncentrat na porost brody i wąsów  Mężczyzna  Zarost   
572                  Booster wzmacniający brodę i włosy  Mężczyzna  Zarost   
711         Pomada do stylizacji  wąsów, brody i włosów  Mężczyzna  Zarost   
744           Krem do brody, skóra wrażliwa, alergiczna  Mężczyzna  Zarost   
759                             Pomada do brody i wąsów  Mężczyzna  Zarost   
1185               Żel do brody, z ekstraktem z Guarany  Mężczyzna  Zarost   
1291  Balsam do stylizacji zarostu, pielęgnacyjny, M...  Mężczyzna  Zarost   
1370       Olejek do brody, Olejek arganowy i Bazyliowy  Mężczyzna  Zarost   
1582  Szampon i odżywka do brody 2w1, Olejek arganow...  Mężczyzna  Zarost   

                     cat_3 cat_4 cat_5  \
236   Akcesoria do zarostu               
325    Pielęgnacja zarostu               
418    Pielęgnacja zarostu               
572    Pielęgnacja zarostu               
711     Stylizacja zarostu               
744     Stylizacja zarostu               
759     Stylizacja zarostu               
1185   Pielęgnacja zarostu               
1291    Stylizacja zarostu               
1370   Pielęgnacja zarostu               
1582     Szampony do brody               

                                                    url  current_price_pln  \
236   https://www.rossmann.pl/Produkt/Akcesoria-do-z...              16.99   
325   https://www.rossmann.pl/Produkt/Pielegnacja-za...              27.99   
418   https://www.rossmann.pl/Produkt/Pielegnacja-za...              33.99   
572   https://www.rossmann.pl/Produkt/Pielegnacja-za...              31.99   
711   https://www.rossmann.pl/Produkt/Stylizacja-zar...              37.99   
744   https://www.rossmann.pl/Produkt/Stylizacja-zar...              32.99   
759   https://www.rossmann.pl/Produkt/Stylizacja-zar...              49.99   
1185  https://www.rossmann.pl/Produkt/Pielegnacja-za...              19.99   
1291  https://www.rossmann.pl/Produkt/Stylizacja-zar...              29.99   
1370  https://www.rossmann.pl/Produkt/Pielegnacja-za...              74.99   
1582  https://www.rossmann.pl/Produkt/Szampony-do-br...              37.99   

      ...  lowest_price_last_30_days_pln  mega_badge  \
236   ...                            NaN       False   
325   ...                            NaN       False   
418   ...                            NaN       False   
572   ...                            NaN       False   
711   ...                            NaN       False   
744   ...                            NaN       False   
759   ...                            NaN       False   
1185  ...                            NaN       False   
1291  ...                            NaN       False   
1370  ...                          93.99        True   
1582  ...                          47.99        True   

                                            ingredients rating  reviews  \
236                        Drewno bukowe, włosia dzika.    5.0      5.0   
325   Aqua, Cera Alba, Isopropyl Palmitate, Copernic...    5.0      2.0   
418   Aqua, Dimethicone, Glycerin, Pentylene Glycol,...    5.0      4.0   
572   Aqua, Alcohol Denat., PEG-40 Hydrogenated Cast...    5.0      1.0   
711   Ingredie

In [65]:
# Usunięcie produktu 2085517 (Szczotka do brody)
df = df.query('product_id != 2085517')
df.query('cat_1 == "Mężczyzna" and cat_2 == "Zarost"')


product_id                    name  \
325       270396             ISANA MEN     
418      2078192           AA MEN Beard    
572      2085518               BARBERO     
711       339519               BARBERO     
744       267581    TOŁPA Dermo Barber.    
759       419910  HAGI Men Whisky Barber   
1185      270391             ISANA MEN     
1291      396104    ARGANOVE Cool Exotic   
1370     2100676       RITUALS... Homme    
1582     2100671       RITUALS... Homme    

                                                caption      cat_1   cat_2  \
325   Pomada do brody, z Olejem Arganowym i Woskiem ...  Mężczyzna  Zarost   
418                  Koncentrat na porost brody i wąsów  Mężczyzna  Zarost   
572                  Booster wzmacniający brodę i włosy  Mężczyzna  Zarost   
711         Pomada do stylizacji  wąsów, brody i włosów  Mężczyzna  Zarost   
744           Krem do brody, skóra wrażliwa, alergiczna  Mężczyzna  Zarost   
759                             Pomada do brody i wąsów  Mężczyzna  Zarost   
1185               Żel do brody, z ekstraktem z Guarany  Mężczyzna  Zarost   
1291  Balsam do stylizacji zarostu, pielęgnacyjny, M...  Mężczyzna  Zarost   
1370       Olejek do brody, Olejek arganowy i Bazyliowy  Mężczyzna  Zarost   
1582  Szampon i odżywka do brody 2w1, Olejek arganow...  Mężczyzna  Zarost   

                    cat_3 cat_4 cat_5  \
325   Pielęgnacja zarostu               
418   Pielęgnacja zarostu               
572   Pielęgnacja zarostu               
711    Stylizacja zarostu               
744    Stylizacja zarostu               
759    Stylizacja zarostu               
1185  Pielęgnacja zarostu               
1291   Stylizacja zarostu               
1370  Pielęgnacja zarostu               
1582    Szampony do brody               

                                                    url  current_price_pln  \
325   https://www.rossmann.pl/Produkt/Pielegnacja-za...              27.99   
418   https://www.rossmann.pl/Produkt/Pielegnacja-za...              33.99   
572   https://www.rossmann.pl/Produkt/Pielegnacja-za...              31.99   
711   https://www.rossmann.pl/Produkt/Stylizacja-zar...              37.99   
744   https://www.rossmann.pl/Produkt/Stylizacja-zar...              32.99   
759   https://www.rossmann.pl/Produkt/Stylizacja-zar...              49.99   
1185  https://www.rossmann.pl/Produkt/Pielegnacja-za...              19.99   
1291  https://www.rossmann.pl/Produkt/Stylizacja-zar...              29.99   
1370  https://www.rossmann.pl/Produkt/Pielegnacja-za...              74.99   
1582  https://www.rossmann.pl/Produkt/Szampony-do-br...              37.99   

      ...  lowest_price_last_30_days_pln  mega_badge  \
325   ...                            NaN       False   
418   ...                            NaN       False   
572   ...                            NaN       False   
711   ...                            NaN       False   
744   ...                            NaN       False   
759   ...                            NaN       False   
1185  ...                            NaN       False   
1291  ...                            NaN       False   
1370  ...                          93.99        True   
1582  ...                          47.99        True   

                                            ingredients rating  reviews  \
325   Aqua, Cera Alba, Isopropyl Palmitate, Copernic...    5.0      2.0   
418   Aqua, Dimethicone, Glycerin, Pentylene Glycol,...    5.0      4.0   
572   Aqua, Alcohol Denat., PEG-40 Hydrogenated Cast...    5.0      1.0   
711   Ingredients: Aqua, Lanolin Cera, Petrolanum, C...    4.0      1.0   
744   Ingredients: Aqua, Propylene Glycol, Glycerin,...    5.0      2.0   
759   Ingredients: Brassica Campestris (Rapeseed) Se...    4.0      1.0   
1185  Ingredients: Aqua, Glycerin, Dicaprylyl Ether,...    5.0      4.0   
1291  Ingredients: INCI: Prunus Amygdalus Dulcis Oil...    4.5      6.0   
1370  Ingredients: Prunus Armeniaca (Apricot) Kernel...    

In [66]:
# Przegląd podkategorii Mężczyzna > Włosy
df.query('cat_1 == "Mężczyzna" and cat_2 == "Włosy"')['cat_3'].value_counts()

cat_3
Szampony i odżywki              16
Koloryzacja włosów i zarostu     9
Stylizacja włosów                4
Akcesoria do włosów              1
Pielęgnacja skóry głowy          1
Name: count, dtype: Int64

In [67]:
# Usunięcie kategorii Akcesoria do włosów
df = df.query('not(cat_1 == "Mężczyzna" and cat_2 == "Włosy" and cat_3 == "Akcesoria do włosów")')
df.query('cat_1 == "Mężczyzna" and cat_2 == "Włosy"')['cat_3'].value_counts()

cat_3
Szampony i odżywki              16
Koloryzacja włosów i zarostu     9
Stylizacja włosów                4
Pielęgnacja skóry głowy          1
Name: count, dtype: Int64

In [68]:
# Przegląd podkategorii Mężczyzna > Higiena
df.query('cat_1 == "Mężczyzna" and cat_2 == "Higiena"')['cat_3'].value_counts()

cat_3
Dezodoranty i antyperspiranty    60
Żele pod prysznic                45
Wkładki higienicze                1
Płyny do higieny intymnej         1
Name: count, dtype: Int64

In [69]:
# Usunięcie kategorii Wkładki higienicze
df = df.query('not(cat_1 == "Mężczyzna" and cat_2 == "Higiena" and cat_3 == "Wkładki higienicze")')
df.query('cat_1 == "Mężczyzna" and cat_2 == "Higiena"')['cat_3'].value_counts()

cat_3
Dezodoranty i antyperspiranty    60
Żele pod prysznic                45
Płyny do higieny intymnej         1
Name: count, dtype: Int64

In [70]:
# Przegląd podkategorii Mężczyzna > Pielęgnacja
df.query('cat_1 == "Mężczyzna" and cat_2 == "Pielęgnacja"')['cat_3'].value_counts()

cat_3
Pielęgnacja twarzy     70
Oczyszczanie twarzy    20
Name: count, dtype: Int64

In [71]:
# Przegląd produktów w kategorii Mężczyzna > Pielęgnacja, które są na sztuki
df.query('cat_1 == "Mężczyzna" and cat_2 == "Pielęgnacja" and capacity_unit == "szt"')

product_id                  name  \
182      2096264     RITUALS... Homme    
290      2094631           ISANA MEN     
866      2098656  NUXE Men Exclusively   
1142     2094630           ISANA MEN     

                                                caption      cat_1  \
182   Zestaw zapach do samochodu + sportowy żel pod ...  Mężczyzna   
290    Płatki pod oczy, hydrożelowe, nawilżająco-kojące  Mężczyzna   
866   Zestaw kosmetyków dla mężczyzn, żel nawilżając...  Mężczyzna   
1142  Płatki pod oczy, hydrożelowe, z Kwasem Hialuro...  Mężczyzna   

            cat_2                cat_3 cat_4 cat_5  \
182   Pielęgnacja  Oczyszczanie twarzy               
290   Pielęgnacja   Pielęgnacja twarzy               
866   Pielęgnacja   Pielęgnacja twarzy               
1142  Pielęgnacja   Pielęgnacja twarzy               

                                                    url  current_price_pln  \
182   https://www.rossmann.pl/Produkt/Oczyszczanie-t...             187.99   
290   https://www.rossmann.pl/Produkt/Pielegnacja-tw...               4.59   
866   https://www.rossmann.pl/Produkt/Pielegnacja-tw...             149.99   
1142  https://www.rossmann.pl/Produkt/Pielegnacja-tw...               4.59   

      ...  lowest_price_last_30_days_pln  mega_badge  \
182   ...                         234.99        True   
290   ...                            NaN       False   
866   ...                            NaN        True   
1142  ...                            NaN       False   

                                            ingredients rating  reviews  \
182                                                        NaN      NaN   
290   Ingredients: Aqua, Glycerin, Sorbitol, Trehalo...    4.7     20.0   
866   Ingredients: Żel pod prysznic: Lauryl Glucose ...    NaN      NaN   
1142  Ingredients: Aqua, Glycerin, Sorbitol, Trehalo...    4.8     19.0   

      capacity_value  capacity_unit price_reference_value  \
182                1            szt                     1   
290                2            szt                     1   
866                1            szt                     1   
1142               2            szt                     1   

      price_reference_unit price_reference_pln  
182                    szt              187.99  
290                    szt                2.30  
866                    szt              149.99  
1142                   szt                2.30  

[4 rows x 21 columns]

In [72]:
# Przegląd podkategorii Mężczyzna > Golenie
df.query('cat_1 == "Mężczyzna" and cat_2 == "Golenie"')['cat_3'].value_counts()

cat_3
Produkty po goleniu                30
Pianki, żele i kremy do golenia     6
Maszynki i trymery do golenia       5
Name: count, dtype: Int64

In [73]:
# Usunięcie kategorii maszynki i trymery do golenia
df = df.query('not(cat_1 == "Mężczyzna" and cat_2 == "Golenie" and cat_3 == "Maszynki i trymery do golenia")')
df.query('cat_1 == "Mężczyzna" and cat_2 == "Golenie"')['cat_3'].value_counts()

cat_3
Produkty po goleniu                30
Pianki, żele i kremy do golenia     6
Name: count, dtype: Int64

In [74]:
# Kategorie pierwszego poziomu
df['cat_1'].value_counts()

cat_1
Mężczyzna                777
Pielęgnacja i higiena    367
Perfumy                  351
Włosy                     82
Lifestyle                  6
Name: count, dtype: Int64

In [75]:
# Na tym etapie mamy wyczyszczone kategorie Mężczyzna i Lifestyle
# Do wyczyszczenia pozostały kategorie "Pielęgnacja i higiena", "Perfumy" i "Włosy"
# Zaczynam od weryfikacji kategorii Perfumy
df[df['cat_1'] == 'Perfumy']['cat_3'].value_counts()

cat_3
Wody toaletowe             269
Wody perfumowane            50
Dezodoranty perfumowane     19
Zestawy zapachowe           13
Name: count, dtype: Int64

In [76]:
df[df['cat_3'] == 'Zestawy zapachowe']

product_id                                   name  \
2        2083791                    JAGUAR Era Reloaded   
14       2086191             POLICE To Be Born To Shine   
70       2087062                   HUGO BOSS The Scent    
85       2084303            JEAN PAUL GAULTIER Le Male    
117      2088899         CRISTIANO RONALDO CR7 Fearless   
127      2105109           CAROLINA HERRERA 212 Men NYC   
150      2074527                    HUGO BOSS Hugo Man    
172      2099189               BVLGARI Man Wood Essence   
184      2092169  DOLCE & GABBANA K by Dolce & Gabbana    
186      2074529                   HUGO BOSS The Scent    
219      2074505                     CALVIN KLEIN Defy    
231      2087061                     HUGO BOSS Bottled    
244       411897                PEPE JEANS Life Is Now    
246      2095983                  PACO RABANNE Phantom    
296      2074530                   HUGO BOSS The Scent    
316      2076958                        LATTAFA Raghba    
318       319247                   JAGUAR Classic Black   
366      2081539                     ICEBERG Twice Gold   
381      2105112                           KENZO Homme    
395      2095989                PACO RABANNE 1 Million    
404      2092518         MONT BLANC Explorer Ultra Blue   
458      2087043                PACO RABANNE 1 Million    
545      2074526             HUGO BOSS Bottled Infinite   
553      2090668                                COACH     
577      2074525                     HUGO BOSS Bottled    
629      2095984                PACO RABANNE 1 Million    
650      2090619                     MONT BLANC Legend    
708      2103934                   DAVIDOFF Cool Water    
728      2069431                         AZZARO Chrome    
751      2074507                 CALVIN KLEIN Eternity    
829      2101154                     HUGO BOSS Bottled    
831      2090620         MONT BLANC Explorer Ultra Blue   
889      2090670                           COACH Green    
933      2087042                  PACO RABANNE Phantom    
939      2087106           GIORGIO ARMANI Acqua di Gio    
967      2087051                     HUGO BOSS Bottled    
991      2087060                     HUGO BOSS Bottled    
1006     2095979                     HUGO BOSS Bottled    
1028     2087044                 PACO RABANNE Invictus    
1060     2084319            JEAN PAUL GAULTIER Scandal    
1162     2090667                      MOSCHINO Toy Boy    
1175     2074501                          GUCCI Guilty    
1236      389667                   LACOSTE Match Point    
1241     2069432                 DIESEL Only The Brave    
1322     2092168              CAROLINA HERRERA Bad Boy    
1334     2074500                 LACOSTE L.12.12 Blanc    
1341     2081563                     HUGO BOSS Bottled    
1376     2103933                CALVIN KLEIN Obsession    
1494     2099188            BVLGARI Man Glacial Essence   
1541     2097856                     HUGO BOSS Bottled    
1573     2095980                  PACO RABANNE Phantom    
1597     2102983                   GUESS Seductive Noir   
1603     2095978                PACO RABANNE 1 Million    
1606     2074498                            HERMES H24    
1632     2074531                   HUGO BOSS The Scent    
1645     2074528                    HUGO BOSS Hugo Man    
1710     2084429            DOLCE & GABBANA Light Blue    
1753     2103935                   DAVIDOFF Cool Water    
1760     2105108              CAROLINA HERRERA Bad Boy    
1774     2069429               BUGATTI Signature  Black   

                                                caption      cat_1  \
2     Zestaw dla mężczyzn, woda toaletowa 100ml + że...  Mężczyzna   
14    Zestaw dla mężczyzn, woda toaletowa 40ml + żel...  Mężczyzna   
70    Zestaw dla mężczyzn, woda toaletowa 100 ml + d...  Mężczyzna   
85    Zestaw dla mężczyzn, woda toaletowa 125ml + ba...  Mężczyzna   
117   Zestaw dla mężczyzn, woda toaletowa 30ml + żel... 

In [77]:
# Na tym etapie zostawiam Zestawy zapachowe. Następną kategorią do weryfikacji jest
# kategoria Włosy
df[df['cat_1'] == 'Włosy']['cat_2'].value_counts()

cat_2
Stylizacja włosów               34
Mycie włosów                    26
Skóra głowy                     16
Koloryzacja włosów               4
Akcesoria i ozdoby do włosów     1
Pielęgnacja włosów               1
Name: count, dtype: Int64

In [78]:
# Usuwam produkty z kategorii Akcesoria i ozdoby do włosów
df = df.query('not(cat_1 == "Włosy" and cat_2 == "Akcesoria i ozdoby do włosów")')
df[df['cat_1'] == 'Włosy']['cat_2'].value_counts()

cat_2
Stylizacja włosów     34
Mycie włosów          26
Skóra głowy           16
Koloryzacja włosów     4
Pielęgnacja włosów     1
Name: count, dtype: Int64

In [79]:
# Weryfikacja kategorii Pielęgnacja i higiena
df[df['cat_1'] == 'Pielęgnacja i higiena']['cat_2'].value_counts()

cat_2
Ciało             203
Broda i zarost    124
Dłonie i stopy     23
Twarz              10
Strefa intymna      6
Senior              1
Name: count, dtype: Int64

In [80]:
# Produkty z powyższej kategorii
df.query('cat_1 == "Pielęgnacja i higiena" and capacity_unit == "szt"')['cat_3'].value_counts()

cat_3
Golenie zarostu                 73
Golenie i depilacja ciała       15
Nietrzymanie moczu               7
Akcesoria do brody i zarostu     5
Zestawy kosmetyków               5
Name: count, dtype: Int64

In [81]:
# Usunięcie produktów powyższych kategorii poza kategorią "Zestawy kosmetyków"
df = df.query('not(cat_1 == "Pielęgnacja i higiena" and capacity_unit == "szt" and cat_3 != "Zestawy kosmetyków")')
df.query('cat_1 == "Pielęgnacja i higiena" and capacity_unit == "szt"')['cat_3'].value_counts()

cat_3
Zestawy kosmetyków    5
Name: count, dtype: Int64

In [82]:
# Przejrzenie kategorii wszystkich pozostałych produktów na sztuki
df.query('capacity_unit == "szt"')['cat_3'].value_counts()

cat_3
Zestawy zapachowe                 58
Koloryzacja włosów i zarostu       6
Zestawy kosmetyków                 5
Serum i wcierki do skóry głowy     4
Pielęgnacja twarzy                 3
Farby do włosów                    3
Oczyszczanie twarzy                1
Pielęgnacja skóry głowy            1
Name: count, dtype: Int64

In [83]:
# Aktualne statystyki df
df.shape

(1482, 21)

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1482 entries, 0 to 1793
Data columns (total 21 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   product_id                     1482 non-null   int64  
 1   name                           1482 non-null   string 
 2   caption                        1482 non-null   string 
 3   cat_1                          1482 non-null   string 
 4   cat_2                          1482 non-null   string 
 5   cat_3                          1482 non-null   string 
 6   cat_4                          1482 non-null   string 
 7   cat_5                          1482 non-null   string 
 8   url                            1482 non-null   string 
 9   current_price_pln              1482 non-null   float64
 10  regular_price_pln              146 non-null    float64
 11  lowest_price_last_30_days_pln  433 non-null    float64
 12  mega_badge                     1482 non-null   bool  

In [85]:
# Zapisanie przeprocesowanych danych do pliku.
df.to_pickle('../data/male_data_preprocessed.pkl')